src/constraint/spec.ts

```diff
@@ -23,45 +23,50 @@ export class SpecConstraintModel extends AbstractConstraintModel {
     super(specConstraint);
   }
 
-  public satisfy(specM: SpecQueryModel, schema: Schema, opt: QueryConfig) {
-    // TODO: Re-order logic to optimize the "requireAllProperties" check
-
-    if (this.constraint.requireAllProperties) {
-      // TODO: extract as a method and do unit test
-      const hasRequiredPropertyAsEnumSpec = some(this.constraint.properties,
-        (prop) => {
-          switch(prop) {
-            // Mark
-            case Property.MARK:
-              return isEnumSpec(specM.getMark());
-
-            // TODO: transform
-
-            // Encoding properties
-            case Property.CHANNEL:
-            case Property.AGGREGATE:
-            case Property.AUTOCOUNT:
-            case Property.BIN:
-            case Property.SCALE:
-            case Property.TIMEUNIT:
-            case Property.FIELD:
-            case Property.TYPE:
-              // If there is property that is enumSpec, we return true as
-              // we cannot check the constraint yet!
-              return some(specM.getEncodings(), (encQ) => {
-                return isEnumSpec(encQ[prop]);
-              });
-            default:
-              /* istanbul ignore next */
-              throw new Error('Unimplemented');
-          }
+    public hasAllRequiredPropertiesSpecific(specM: SpecQueryModel): boolean {
+      return every(this.constraint.properties, (prop) => {
+        if (prop === Property.MARK) {
+          return !isEnumSpec(specM.getMark());
+        }
+
+        // TODO: transform
+
+        const nestedEncProp = getNestedEncodingProperty(prop);
+
+        if (nestedEncProp) {
+          let parent = nestedEncProp.parent;
+          let child = nestedEncProp.child;
+
+          return every(specM.getEncodings(), (encQ) => {
+            if (!encQ[parent]) {
+              return true;
+            }
+
+            return !isEnumSpec(encQ[parent][child]);
+          });
         }
-      );
-      // If one of the required property is still an enum spec, do not check the constraint yet.
-      if (hasRequiredPropertyAsEnumSpec) {
-        return true; // Return true since the query still satisfy the constraint.
+
+        if (!contains(ENCODING_PROPERTIES, prop)) {
```

Comment:
This takes `O(N)` to just check for error every time and this method get called very often.  Do you think this is an appropriate implementation?